File: agentic_chatbot_flutter/lib/src/widgets/chat_bubble.dart
import 'package:flutter/material.dart';
import '../models/message.dart';
import '../models/theme.dart';

/// Chat bubble widget for individual messages
class ChatBubble extends StatelessWidget {
  final Message message;
  final AgenticTheme theme;
  final void Function(FeedbackType)? onFeedback;

  const ChatBubble({
    super.key,
    required this.message,
    required this.theme,
    this.onFeedback,
  });

  @override
  Widget build(BuildContext context) {
    final isUser = message.role == MessageRole.user;
    final isError = message.isError;

    return Padding(
      padding: const EdgeInsets.only(bottom: 12),
      child: Align(
        alignment: isUser ? Alignment.centerRight : Alignment.centerLeft,
        child: Column(
          crossAxisAlignment: isUser ? CrossAxisAlignment.end : CrossAxisAlignment.start,
          children: [
            // Bubble
            Container(
              constraints: BoxConstraints(
                maxWidth: MediaQuery.of(context).size.width * 0.8,
              ),
              padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 10),
              decoration: BoxDecoration(
                color: isUser
                    ? theme.userBubbleColor
                    : isError
                        ? const Color(0xFFFEE2E2)
                        : theme.assistantBubbleColor,
                borderRadius: BorderRadius.circular(18),
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.end,
                children: [
                  Text(
                    message.content,
                    style: TextStyle(
                      color: isUser
                          ? theme.userTextColor
                          : isError
                              ? const Color(0xFFDC2626)
                              : theme.assistantTextColor,
                      fontSize: 15,
                      height: 1.4,
                    ),
                  ),
                  const SizedBox(height: 4),
                  Text(
                    _formatTime(message.timestamp),
                    style: TextStyle(
                      color: isUser
                          ? theme.userTextColor.withOpacity(0.7)
                          : const Color(0xFF94A3B8),
                      fontSize: 11,
                    ),
                  ),
                ],
              ),
            ),

            // Feedback buttons
            if (!isUser && !isError && onFeedback != null)
              Padding(
                padding: const EdgeInsets.only(top: 6),
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    _FeedbackButton(
                      icon: '👍',
                      isActive: message.feedback == FeedbackType.positive,
                      onTap: () => onFeedback!(FeedbackType.positive),
                    ),
                    const SizedBox(width: 8),
                    _FeedbackButton(
                      icon: '👎',
                      isActive: message.feedback == FeedbackType.negative,
                      onTap: () => onFeedback!(FeedbackType.negative),
                    ),
                  ],
                ),
              ),
          ],
        ),
      ),
    );
  }

  String _formatTime(DateTime date) {
    final hours = date.hour.toString().padLeft(2, '0');
    final minutes = date.minute.toString().padLeft(2, '0');
    return '$hours:$minutes';
  }
}

class _FeedbackButton extends StatelessWidget {
  final String icon;
  final bool isActive;
  final VoidCallback onTap;

  const _FeedbackButton({
    required this.icon,
    required this.isActive,
    required this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.all(4),
        decoration: BoxDecoration(
          color: isActive ? const Color(0xFFDBEAFE) : const Color(0xFFF1F5F9),
          borderRadius: BorderRadius.circular(12),
        ),
        child: Text(icon, style: const TextStyle(fontSize: 14)),
      ),
    );
  }
}

File: react-native-agentic-chatbot/src/components/ChatBubble.js
/**
 * ChatBubble Component
 * Message bubble component for React Native
 */

import React from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';

/**
 * ChatBubble - Individual message bubble
 */
export function ChatBubble({ message, colors, onFeedback }) {
    const isUser = message.role === 'user';
    const isError = message.isError;

    return (
        <View style={[
            styles.container,
            isUser ? styles.userContainer : styles.assistantContainer,
        ]}>
            <View style={[
                styles.bubble,
                isUser
                    ? { backgroundColor: colors.userBubbleColor }
                    : { backgroundColor: isError ? '#fee2e2' : colors.assistantBubbleColor },
            ]}>
                <Text style={[
                    styles.text,
                    isUser
                        ? { color: colors.userTextColor }
                        : { color: isError ? '#dc2626' : colors.assistantTextColor },
                ]}>
                    {message.content}
                </Text>

                {/* Timestamp */}
                <Text style={[styles.timestamp, { color: isUser ? 'rgba(255,255,255,0.7)' : '#94a3b8' }]}>
                    {formatTime(message.timestamp)}
                </Text>
            </View>

            {/* Feedback buttons (only for assistant messages) */}
            {!isUser && !isError && onFeedback && (
                <View style={styles.feedbackContainer}>
                    <TouchableOpacity
                        style={[
                            styles.feedbackButton,
                            message.feedback === 'positive' && styles.feedbackActive,
                        ]}
                        onPress={() => onFeedback('positive')}
                    >
                        <Text style={styles.feedbackIcon}>👍</Text>
                    </TouchableOpacity>
                    <TouchableOpacity
                        style={[
                            styles.feedbackButton,
                            message.feedback === 'negative' && styles.feedbackActive,
                        ]}
                        onPress={() => onFeedback('negative')}
                    >
                        <Text style={styles.feedbackIcon}>👎</Text>
                    </TouchableOpacity>
                </View>
            )}
        </View>
    );
}

/**
 * Format timestamp
 */
function formatTime(date) {
    if (!date) return '';
    const d = new Date(date);
    return d.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
}

const styles = StyleSheet.create({
    container: {
        marginBottom: 12,
        maxWidth: '80%',
    },
    userContainer: {
        alignSelf: 'flex-end',
    },
    assistantContainer: {
        alignSelf: 'flex-start',
    },
    bubble: {
        paddingHorizontal: 14,
        paddingVertical: 10,
        borderRadius: 18,
    },
    text: {
        fontSize: 15,
        lineHeight: 21,
    },
    timestamp: {
        fontSize: 11,
        marginTop: 4,
        textAlign: 'right',
    },
    feedbackContainer: {
        flexDirection: 'row',
        marginTop: 6,
        gap: 8,
    },
    feedbackButton: {
        padding: 4,
        borderRadius: 12,
        backgroundColor: '#f1f5f9',
    },
    feedbackActive: {
        backgroundColor: '#dbeafe',
    },
    feedbackIcon: {
        fontSize: 14,
    },
});
